sys/linux/rdma.txt

```diff
@@ -0,0 +1,1464 @@
+ib_read_counters_flags = IB_UVERBS_READ_COUNTERS_PREFER_CACHED
+resource vcontext_handle[int32]
+
+ib_uverbs_create_counters_cmd {
+	length		bytesize8[ib_uverbs_create_counters_cmd, int16]
+	object_id	const[UVERBS_OBJECT_COUNTERS, int16]
+	method_id	const[UVERBS_METHOD_COUNTERS_CREATE, int16]
+	num_attrs	const[1, int16]
+	reserved0	const[0x0, int64]
+	driver_id	const[RDMA_DRIVER_MLX5, int32]
+	reserved1	const[0x0, int32]
+
+	attr_id		const[UVERBS_ATTR_CREATE_COUNTERS_HANDLE, int16]
+	len		int16
+	flags		flags[ib_uverbs_attr_flags, int16]
+	elem_id		int8
+	reserved2	const[0x0, int8]
+	reserved3	const[0x0, int16]
+	vcontext_handle	vcontext_handle[opt]
+	more_data	int32
+}
+
+ib_uverbs_destroy_counters_cmd {
+	length		bytesize8[ib_uverbs_destroy_counters_cmd, int16]
+	object_id	const[UVERBS_OBJECT_COUNTERS, int16]
+	method_id	const[UVERBS_METHOD_COUNTERS_DESTROY, int16]
+	num_attrs	const[1, int16]
+	reserved0	const[0x0, int64]
+	driver_id	const[RDMA_DRIVER_MLX5, int32]
+	reserved1	const[0x0, int32]
+
+	attr_id		const[UVERBS_ATTR_DESTROY_COUNTERS_HANDLE, int16]
+	len		int16
+	flags		flags[ib_uverbs_attr_flags, int16]
+	elem_id		int8
+	reserved2	const[0x0, int8]
+	reserved3	const[0x0, int16]
+	vcontext_handle	vcontext_handle
+	more_data	int32
+}
+
+# in UVERBS_ATTR_READ_COUNTERS_FLAGS, assuming that sizeof(uint32_t is 4)
+ib_uverbs_read_counters_cmd {
+	length		bytesize8[ib_uverbs_read_counters_cmd, int16]
+	object_id	const[UVERBS_OBJECT_COUNTERS, int16]
+	method_id	const[UVERBS_METHOD_COUNTERS_READ, int16]
+	num_attrs	const[0x3, int16]
+	reserved0	const[0x0, int64]
+	driver_id	const[RDMA_DRIVER_MLX5, int32]
+	reserved1	const[0x0, int32]
+
+	attr_id0	const[UVERBS_ATTR_READ_COUNTERS_HANDLE, int16]
+	len0		int16
+	flags0		flags[ib_uverbs_attr_flags, int16]
+	reserved2	const[0x0, int16]
+	vcontext_handle	vcontext_handle
+	more_data0	int32
+
+	attr_id1	const[UVERBS_ATTR_READ_COUNTERS_BUFF, int16]
+	len1		bytesize8[data, int16]
+	flags1		flags[ib_uverbs_attr_flags, int16]
+	reserved3	const[0x0, int16]
+	data		buffer[out]
+
+	attr_id2	const[UVERBS_ATTR_READ_COUNTERS_FLAGS, int16]
+	len2		const[0x4, int16]
+	flags2		flags[ib_uverbs_attr_flags, int16]
+	reserved4	const[0x0, int16]
+	flags		flags[ib_read_counters_flags, int32]
+	more_data1	int32
+}
+
+# commands
+ioctl$CREATE_COUNTERS(fd fd_rdma, cmd const[RDMA_VERBS_IOCTL], arg ptr[inout, ib_uverbs_create_counters_cmd])
+ioctl$DESTROY_COUNTERS(fd fd_rdma, cmd const[RDMA_VERBS_IOCTL], arg ptr[in, ib_uverbs_destroy_counters_cmd])
+ioctl$READ_COUNTERS(fd fd_rdma, cmd const[RDMA_VERBS_IOCTL], arg ptr[inout, ib_uverbs_read_counters_cmd])
+
+# device
+syz_open_dev$ibv_device(dev ptr[in, string["/dev/infiniband/uverbs0"]], id intptr, flags flags[rdma_dev_open_flags]) fd_rdma
+write$MLX5_GET_CONTEXT(fd fd_rdma, buf ptr[inout, mlx5_get_context_cmd], len len[buf])
+close$ibv_device(fd fd_rdma)
+write$QUERY_DEVICE_EX(fd fd_rdma, buf ptr[inout, query_device_cmd_ex], len len[buf])
```

Comment:
The query_device_cmd_ex struct itself is only input, right?
There is ptr[out] for response and that response is output, right?
If so, use `in` direction here and for other write's. ptr[out] for response will override it for the response, so it still will be output.